### ISSUE TITLE
Listing options cause an error and change device settings

### ISSUE THREAD

**Issue by ifohancroft** [LoLei/razer-cli#65]
Ideally, this should probably be two different bug reports and I should test if the other options reset the devices' settings but right now I don't feel like having to reboot into Windows to fix my device's settings after.

Basically, when executing razer-cli -l or razer-cli -ll my Viper Ultimate's DPI gets set to 1800 and the polling rate to 500 (according to razer-cli's output and I get the following error/output:

Found 2 Razer device(s)
Razer Viper Ultimate (Wireless):
   type: mouse
   DPI: (1800, 1800)
   max DPI: 20000
   polling rate: 500
   battery:
      charge: 75
      charging: False
Traceback (most recent call last):
  File "/usr/bin/razer-cli", line 33, in <module>
    sys.exit(load_entry_point('razer-cli==2.0.1', 'console_scripts', 'razer-cli')())
  File "/usr/lib/python3.9/site-packages/razer_cli-2.0.1-py3.9.egg/razer_cli/razer_cli/main.py", line 30, in main
    rc.run()
  File "/usr/lib/python3.9/site-packages/razer_cli-2.0.1-py3.9.egg/razer_cli/razer_cli/rc.py", line 73, in run
    self.device_lister.list()
  File "/usr/lib/python3.9/site-packages/razer_cli-2.0.1-py3.9.egg/razer_cli/razer_cli/lister/device_lister.py", line 50, in list
    device.get_low_battery_threshold(), '%')
  File "/usr/lib/python3.9/site-packages/openrazer/client/devices/mice.py", line 281, in get_low_battery_threshold
    return int(self._dbus_interfaces['power'].getLowBatteryThreshold())
  File "/usr/lib/python3.9/site-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "/usr/lib/python3.9/site-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/dbus/service.py", line 658, in _message_cb
    (candidate_method, parent_method) = _method_lookup(self, method_name, interface_name)
  File "/usr/lib/python3.9/site-packages/dbus/service.py", line 248, in _method_lookup
    raise UnknownMethodException('%s is not a valid method of interface %s' % (method_name, dbus_interface))
dbus.exceptions.UnknownMethodException: org.freedesktop.DBus.Error.UnknownMethod: Unknown method: getLowBatteryThreshold is not a valid method of interface razer.device.power

I have my mouse's dongle plugged into the dock and the dock plugged into the PC. I also have my Razer Kiyo webcam plugged into the PC but since the Kiyo doesn't work with openrazer, I am guessing the 2 devices razer-cli sees are the Mouse and Dock.

I am using razer-cli version 3.0.1 from GitHub and apparently even checking the version messes up the device's settings.

**Comment by ifohancroft** [comment]
To clarify, when I say I am using version 3.0.1. I don't mean from the releases (I don't know if there is such a release). I cloned the repo and installed it with ```sudo python setup.py install``` version 3.0.1. is what razer-cli's --version option gives me as the version number.

**Comment by LoLei** [comment]
Thanks for the report!

I tried reproducing this. I don't have the same hardware so out of the box this does not happen for me. I've also tried hardcoding `if device.has('battery'):` in `devicer_lister.py` to true, and even manually throwing an error within there. (This is the same place where your exception happens.) I assume the error happens because while `device.has('battery')` may be true, the actual dbus method `getLowBatteryThreshold` does not exist. This could be fixed with a try/catch.

None of this changes my DPI or polling settings.

Some questions:
1. You said `--version` alone also changes your settings. Does an error happen there as well?
2. Could you please re-run these commands (`-l`, `--version`, whichever change your settings) with the verbose (`-v`) flag? This should produce some output that is called when the DPI and/or polling rate is changed via razer-cli's `dpi-setter.py` and `poll-rater-setter.py`. I'd like to see those logs.
3. I assume you mean razer-cli version 2.0.1? That is the latest tag. 3.0.1 is the latest version of openrazer (which also shows up in `--version`).

**Comment by ifohancroft** [comment]
>     1. You said `--version` alone also changes your settings. Does an error happen there as well?

No. I just ran ```razer-cli -v --version``` and while it did change the settings, I did not get an error. Should I also try razer-cli --version on its own to check if it produces any errors?
 
>     2. Could you please re-run these commands (`-l`, `--version`, whichever change your settings) with the verbose (`-v`) flag? This should produce some output that is called when the DPI and/or polling rate is changed via razer-cli's `dpi-setter.py` and `poll-rater-setter.py`. I'd like to see those logs.

Just did, but there isn't much output. This is all it shows:

```
razer-cli: 2.0.1
   Installed in: /usr/lib/python3.9/site-packages/razer_cli-2.0.1-py3.9.egg/razer_cli
python3-openrazer: 3.0.1
openrazer-daemon: 3.0.1
Python: 3.9.5
```

>     3. I assume you mean razer-cli version 2.0.1? That is the latest tag. 3.0.1 is the latest version of openrazer (which also shows up in `--version`).

Yes, sorry, it's 2.0.1

**Comment by LoLei** [comment]
Running with `-v` and not getting any additional output should confirm that the DPI and polling rate are not changed by razer-cli code, otherwise it would show up in the logs:
* https://github.com/LoLei/razer-cli/blob/master/razer_cli/razer_cli/setter/dpi_setter.py#L30
* https://github.com/LoLei/razer-cli/blob/master/razer_cli/razer_cli/setter/poll_rate_setter.py#L23

There is also no way the settings are somehow being restored from the cache file, as that feature is incomplete and only prints the needed command for you to restore all settings (and this method should also not even be called by accident):
* https://github.com/LoLei/razer-cli/blob/master/razer_cli/razer_cli/util.py#L86

As a last-ditch effort could you check out this branch: https://github.com/LoLei/razer-cli/tree/test/65
I've commented out everything in the `run` method except the part which handles printing the version. We can try to eliminate possible other culprits this way.

**Comment by ifohancroft** [comment]
Hmm.. weird.

Sure, I wouldn't mind testing. Any idea on how to remove the previous version first? Apparently the setup.py doesn't have either 'remove' nor 'uninstall'.

**Comment by LoLei** [comment]
I'd use pip, which can uninstall stuff installed with setup.py, but if you don't have that anymore I don't know.

I'd also recommend installing it via `pip install -e . --user`, if you're doing testing, as that will automatically use code changes in the installed package without having to re-install it.

**Comment by ifohancroft** [comment]
I installed again with: ```sudo python setup.py install``` but this time using ```--record <filename>``` so the full command became ```sudo pythong setup.py install --record files.txt``` and it listed all installed files, which were just the /usr/bin/razer-cli and a folder called razer_cli-2.0.1-py3.9.egg under /usr/lib/python3.9/site-packages/ so I have removed them.

I have now cloned the https://github.com/LoLei/razer-cli/tree/test/65 branch and have installed it with ```sudo pip install .``` although it told me that it is using legacy setup.py install as package wheel was not installed.

So, I have tested running ```razer-cli -v -l``` and the mouse changes again. Here is the output I get:

Found 2 Razer device(s)
Razer Viper Ultimate (Wireless):
   type: mouse
   DPI: (1800, 1800)
   max DPI: 20000
   polling rate: 500
   battery:
      charge: 40
      charging: False
Traceback (most recent call last):
  File "/usr/bin/razer-cli", line 33, in <module>
    sys.exit(load_entry_point('razer-cli==2.0.1', 'console_scripts', 'razer-cli')())
  File "/usr/lib/python3.9/site-packages/razer_cli/razer_cli/main.py", line 30, in main
    rc.run()
  File "/usr/lib/python3.9/site-packages/razer_cli/razer_cli/rc.py", line 73, in run
    self.device_lister.list()
  File "/usr/lib/python3.9/site-packages/razer_cli/razer_cli/lister/device_lister.py", line 50, in list
    device.get_low_battery_threshold(), '%')
  File "/usr/lib/python3.9/site-packages/openrazer/client/devices/mice.py", line 281, in get_low_battery_threshold
    return int(self._dbus_interfaces['power'].getLowBatteryThreshold())
  File "/usr/lib/python3.9/site-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "/usr/lib/python3.9/site-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/dbus/service.py", line 658, in _message_cb
    (candidate_method, parent_method) = _method_lookup(self, method_name, interface_name)
  File "/usr/lib/python3.9/site-packages/dbus/service.py", line 248, in _method_lookup
    raise UnknownMethodException('%s is not a valid method of interface %s' % (method_name, dbus_interface))
dbus.exceptions.UnknownMethodException: org.freedesktop.DBus.Error.UnknownMethod: Unknown method: getLowBatteryThreshold is not a valid method of interface razer.device.power


P.S. I've been meaning to ask you: To set settings to a specific device, it asks you that you specify the device. How do you specify the device? I mean like do you use a device id or something and where do I get the device's id?

**Comment by LoLei** [comment]
Something with the branch checkout or the installation must not have worked since the device listing is actually commented out on this branch now:

https://github.com/LoLei/razer-cli/blob/test/65/razer_cli/razer_cli/rc.py#L72

I only left the version command in so only `razer-cli -v --version` should work in this branch, `-l` should not return any listing output.

You may need to check out the branch manually, I noticed you only said you cloned the repository. Or you can specify the branch while cloning: `git clone -b test/65 git@github.com:LoLei/razer-cli.git razer-cli-with-branch-checked-out`

>P.S. I did miss some files when uninstalling so I uninstalled again, installed again and tried again but the result was the same. Perhaps, if it's not razer-cli resetting the device, I need to use razer-cli to set the device's DPI to 400 and the polling rate to 1000Hz, then try to start razer-cli again to see if it changes it.

I'm not sure I fully understand. When I tested it I always used razer-cli to change DPI and polling rate, ran `-l` or `--version`, saw that it remained the same. That multiple times. I assumed this is what you did as well.

>P.P.S. I've been meaning to ask you: To set settings to a specific device, it asks you that you specify the device. How do you specify the device? I mean like do you use a device id or something and where do I get the device's id?

This is actually mentioned in `-d`'s helptext (`razer-cli -h`), and `razer-cli -man device`, but I guess it has gotten more confusing since the outputs of the list arguments have gotten so huge. It's simply the exact name of the device as it's mentioned in the output of a list argument, in quotes, e.g. `razer-cli -d "Razer Lancehead Tournament Edition" -e multicolor`, or the device's serial number, which is also in the list output, e.g. `razer-cli -d 341821F05301597 -c ff00ff`.

**Comment by ifohancroft** [comment]
> Something with the branch checkout or the installation must not have worked since the device listing is actually commented out on this branch now:
> 
> https://github.com/LoLei/razer-cli/blob/test/65/razer_cli/razer_cli/rc.py#L72
> 
> I only left the version command in so only `razer-cli -v --version` should work in this branch, `-l` should not return any listing output.
> 
> You may need to check out the branch manually, I noticed you only said you cloned the repository. Or you can specify the branch while cloning: `git clone -b test/65 git@github.com:LoLei/razer-cli.git razer-cli-with-branch-checked-out`
> 
> > P.S. I did miss some files when uninstalling so I uninstalled again, installed again and tried again but the result was the same. Perhaps, if it's not razer-cli resetting the device, I need to use razer-cli to set the device's DPI to 400 and the polling rate to 1000Hz, then try to start razer-cli again to see if it changes it.
> 
> I'm not sure I fully understand. When I tested it I always used razer-cli to change DPI and polling rate, ran `-l` or `--version`, saw that it remained the same. That multiple times. I assumed this is what you did as well.
> 
> > P.P.S. I've been meaning to ask you: To set settings to a specific device, it asks you that you specify the device. How do you specify the device? I mean like do you use a device id or something and where do I get the device's id?
> 
> This is actually mentioned in `-d`'s helptext (`razer-cli -h`), and `razer-cli -man device`, but I guess it has gotten more confusing since the outputs of the list arguments have gotten so huge. It's simply the exact name of the device as it's mentioned in the output of a list argument, in quotes, e.g. `razer-cli -d "Razer Lancehead Tournament Edition" -e multicolor`, or the device's serial number, which is also in the list output, e.g. `razer-cli -d 341821F05301597 -c ff00ff`.

Hmm. Good point! Thank you! I will clone again and check to make sure I am on the right branch before installing.

Oh! That makes sense. Thank you!

Currently, the output only lists the mouse though 'Razer Viper Ultimate (Wireless)' but I don't remember how the output looked before I started getting this error. Is the dock not listed because the error happens before it reaches to it?

**Comment by ifohancroft** [comment]
Ok, the -l does not return output now but it does again changes the device dpi and stuff.

Executing razer-cli -v --version after that, I get:

razer-cli: 2.0.1
   Installed in: /usr/lib/python3.9/site-packages/razer_cli
python3-openrazer: 3.0.1
openrazer-daemon: 3.0.1
Python: 3.9.5

Should I have tested directly with razer-cli -v --version?

As for the DPI: After I execute razer-cli and it changes the DPI and polling rate, I boot into Windows and start Synapse to return the device to normal state then boot back into Linux. In Linux I keep the daemon off as I don't want it constantly running, razer-cli starts it when it needs it and after I am done making changes with razer-cli I just kill the daemon process usually.

**Comment by ifohancroft** [comment]
I just tested with razer-cli -v --version directly (I mean after fixing the DPI) and it resets things again. The output is:

razer-cli: 2.0.1
   Installed in: /usr/lib/python3.9/site-packages/razer_cli
python3-openrazer: 3.0.1
openrazer-daemon: 3.0.1
Python: 3.9.5

**Comment by LoLei** [comment]
Ah, so you're rebooting in between these tests/changes. That should explain something. My bad if I missed that before.

I just noticed that openrazer 3.0.0 introduced some changes regarding persistence on reboot. I just tested it with DPI and poll rate, and that along with some colors and effects is indeed being reset after a reboot. (I just never noticed because I mostly run razer-cli with some settings like `-x` on startup.)

[Release 3.0.0](https://github.com/openrazer/openrazer/releases/tag/v3.0.0) states:
>Add persistent storage of effects in daemon

[Release 3.0.1](https://github.com/openrazer/openrazer/releases/tag/v3.0.1) adds:
>Add config option to disable a new feature in v3.0.0 that restores the lighting effects.
(https://github.com/openrazer/openrazer/commit/406041d1ee730ce513153cc0d759d843c73a0e5f)

The config file information can be read at `man razer.conf`. The specific setting seems to be `restore_persistence`.

So I take it openrazer uses saved settings from `persistance.conf`, which is in the same directory as `razer.conf` (`~/.config/openrazer`).

I just tested setting this flag to false and it did indeed fix the issue, presumably this should also work for you. For reference, here is my config file as it exists now: [`razer.conf`](https://github.com/LoLei/dotfiles/blob/6495fa54525564e413bf24c16c31dbeb360943ca/.config/openrazer/razer.conf).

I don't fully understand the purpose of the persistance storage and how or when the `persistance.conf` file is being written, but since that behavior can be disabled via the config setting I don't really care to look into it right now. For the future it might however be cool if razer-cli uses that file as well.

**Comment by ifohancroft** [comment]
Thank you! I'd look into that.
That explains a lot! It's weird that they set the file to have default values from the start, which suck as it will reset/change your device the first time. Also, it's weird there is a need for that file since changes from openrazer and razer-cli have always been persistent.

P.S. I feel like razer-cli having its own file makes more sense, unless openrazer's persistance files interferes.

**Comment by ifohancroft** [comment]
Just tested the whole persistance thing.

It's a shitshow. Not sure why they added it. I deleted all configs possible and started the application (btw the error I get from razer-cli when listing devices still exists) and things got reset again.

I have the whole persistance thing off by default in the config. Perhaps, I can set it to True, then fix the DPI and Polling rate in persistance.conf so my mouse gets returned back to normal.

P.S. Just tested this (Kill the daemon, changed the config to turn on Persistance and changed the DPI and Polling Rate to 400 DPI, 1000Hz, then executed razer-cli and my mouse got returned to normal.

However, I killed the daemon again, turned off persistance, started razer-cli again, so the daemon started and gets the persistance off setting, then killed it again, changed the DPI in the file, started it again, and again it changes my mouse, so I guess the persistance option is only about saving the setting, not about reading it and it always reads it regardless of whether persistance is off. Anyway, I need to report this to openrazer, however, like mentioned earlier the razer-cli error still persists.
